File: templates/zig/src/plugins/llamaindex.zig
//! LlamaIndex → Borgkit Plugin (Zig) — HTTP Bridge
//!
//! Wraps a LlamaIndex chat/query server so it participates in the Borgkit mesh
//! as a standard IAgent.  Uses `std.http.Client` for all HTTP communication —
//! no external dependencies required.
//!
//! ── LlamaIndex API contract ───────────────────────────────────────────────────
//!
//!   POST {base_url}/chat
//!     Body:     { "message": "...", "chat_history": [] }
//!     Response: { "response": "...", "source_nodes": [] }
//!
//! The server is typically started via LlamaIndex's built-in FastAPI integration:
//!
//! ── Setup ──────────────────────────────────────────────────────────────────────
//!
//!   from llama_index.core import VectorStoreIndex, SimpleDirectoryReader
//!   from llama_index.server import LlamaIndexServer
//!
//!   index  = VectorStoreIndex.from_documents(
//!                SimpleDirectoryReader("data").load_data())
//!   engine = index.as_chat_engine()
//!   server = LlamaIndexServer(chat_engine=engine, port=8080)
//!   server.run()
//!
//!   # Or with llama-index-server:
//!   # llama-index-server --port 8080
//!
//! ── Usage ──────────────────────────────────────────────────────────────────────
//!
//!   const llamaindex = @import("plugins/llamaindex.zig");
//!
//!   var service = llamaindex.LlamaIndexService{
//!       .base_url = "http://localhost:8080",
//!   };
//!   var plugin = llamaindex.LlamaIndexPlugin.init(allocator);
//!   defer plugin.deinit();
//!
//!   const Wrapped = wrapped_agent.WrappedAgent(
//!       llamaindex.LlamaIndexService, llamaindex.LlamaIndexPlugin);
//!   var agent = Wrapped.init(&service, &plugin, .{
//!       .agent_id = "borgkit://agent/llamaindex",
//!       .owner    = "0xYourWallet",
//!   }, allocator);
//!
//!   const resp = agent.handleRequest(req);

const std     = @import("std");
const types   = @import("../types.zig");
const iplugin = @import("iPlugin.zig");

// ── Service config (the "agent" token) ────────────────────────────────────────

/// Configuration for a LlamaIndex HTTP chat endpoint.
pub const LlamaIndexService = struct {
    /// Base URL, e.g. "http://localhost:8080".
    base_url: []const u8 = "http://localhost:8080",

    /// POST path for chat invocation (default: "/chat").
    invoke_route: []const u8 = "/chat",

    /// Optional API key sent as "Authorization: Bearer <key>".
    api_key: ?[]const u8 = null,
};

// ── Plugin ────────────────────────────────────────────────────────────────────

/// HTTP bridge plugin for LlamaIndex chat servers.
///
/// Stores a reusable `std.http.Client`; call `deinit()` when done.
pub const LlamaIndexPlugin = struct {
    allocator: std.mem.Allocator,
    client:    std.http.Client,

    pub fn init(allocator: std.mem.Allocator) LlamaIndexPlugin {
        return .{
            .allocator = allocator,
            .client    = std.http.Client{ .allocator = allocator },
        };
    }

    pub fn deinit(self: *LlamaIndexPlugin) void {
        self.client.deinit();
    }

    // ── IPlugin interface ───────────────────────────────────────────────────

    /// Return a single "invoke" capability — capabilities are static for HTTP
    /// bridge plugins.  Pass explicit capabilities at wrap time via
    /// `WrappedAgent.init` if you need multiple.
    pub fn extractCapabilities(
        _self:     *LlamaIndexPlugin,
        _agent:    *LlamaIndexService,
        allocator: std.mem.Allocator,
    ) []iplugin.CapabilityDescriptor {
        const caps = allocator.alloc(iplugin.CapabilityDescriptor, 1) catch
            return &[_]iplugin.CapabilityDescriptor{};
        caps[0] = .{
            .name        = "invoke",
            .description = "Invoke the LlamaIndex agent via its HTTP chat endpoint",
        };
        return caps;
    }

    /// Build the LlamaIndex chat JSON body:
    ///   { "message": <payload>, "chat_history": [] }
    ///
    /// Extracts `message`, `input`, or `query` key from the request payload if
    /// present; otherwise uses the raw payload string as the message.
    pub fn translateRequest(
        _self:     *LlamaIndexPlugin,
        req:       types.AgentRequest,
        allocator: std.mem.Allocator,
    ) ![]const u8 {
        // Try to extract a string content field from the JSON payload
        const content = blk: {
            const parsed = std.json.parseFromSlice(
                std.json.Value, allocator, req.payload, .{ .ignore_unknown_fields = true },
            ) catch break :blk req.payload;
            defer parsed.deinit();

            if (parsed.value == .object) {
                for ([_][]const u8{ "message", "input", "query" }) |key| {
                    if (parsed.value.object.get(key)) |v| {
                        if (v == .string) {
                            const copy = try allocator.dupe(u8, v.string);
                            break :blk copy;
                        }
                    }
                }
            }
            break :blk try allocator.dupe(u8, req.payload);
        };
        defer allocator.free(content);

        // JSON-encode the content string so it is safe inside the body
        var escaped = std.ArrayList(u8).init(allocator);
        defer escaped.deinit();
        try std.json.encodeJsonString(content, .{}, escaped.writer());

        return std.fmt.allocPrint(allocator,
            \\{{"message":{s},"chat_history":[]}}
            , .{escaped.items});
    }

    /// Extract the agent reply from the LlamaIndex response.
    ///
    /// Looks for the top-level `response` field; falls back to returning the
    /// raw JSON if the field is absent or not a string.
    pub fn translateResponse(
        _self:      *LlamaIndexPlugin,
        request_id: []const u8,
        raw:        []const u8,
    ) types.AgentResponse {
        const content = extractResponseField(raw) orelse raw;
        return types.AgentResponse.success(request_id, content);
    }

    /// POST the translated body to `{base_url}{invoke_route}` and return the
    /// raw response bytes.
    pub fn invokeNative(
        self:      *LlamaIndexPlugin,
        agent:     *LlamaIndexService,
        input:     []const u8,
        allocator: std.mem.Allocator,
    ) ![]const u8 {
        const url = try std.fmt.allocPrint(
            allocator, "{s}{s}",
            .{ trimSlash(agent.base_url), agent.invoke_route },
        );
        defer allocator.free(url);

        return self.postJson(agent, url, input, allocator);
    }

    // ── Internal HTTP helper ────────────────────────────────────────────────

    fn postJson(
        self:      *LlamaIndexPlugin,
        agent:     *LlamaIndexService,
        url:       []const u8,
        payload:   []const u8,
        allocator: std.mem.Allocator,
    ) ![]const u8 {
        var auth_buf: [512]u8 = undefined;
        var hdrs_buf: [3]std.http.Header = undefined;
        var n: usize = 0;
        hdrs_buf[n] = .{ .name = "Content-Type", .value = "application/json" };
        n += 1;
        hdrs_buf[n] = .{ .name = "Accept", .value = "application/json" };
        n += 1;
        if (agent.api_key) |key| {
            const bearer = try std.fmt.bufPrint(&auth_buf, "Bearer {s}", .{key});
            hdrs_buf[n] = .{ .name = "Authorization", .value = bearer };
            n += 1;
        }
        const extra = hdrs_buf[0..n];

        var body = std.ArrayList(u8).init(allocator);
        errdefer body.deinit();

        const fr = try self.client.fetch(.{
            .method           = .POST,
            .location         = .{ .url = url },
            .extra_headers    = extra,
            .payload          = payload,
            .response_storage = .{ .dynamic = &body },
        });

        const code = @intFromEnum(fr.status);
        if (code < 200 or code >= 300) return error.LlamaIndexBadStatus;

        return body.toOwnedSlice();
    }
};

// ── Helpers ───────────────────────────────────────────────────────────────────

/// Parse a LlamaIndex response JSON and return the `response` field value.
/// Returns null if the JSON cannot be parsed or the field is not a string.
fn extractResponseField(raw: []const u8) ?[]const u8 {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const alloc = arena.allocator();

    const parsed = std.json.parseFromSlice(
        std.json.Value, alloc, raw, .{ .ignore_unknown_fields = true },
    ) catch return null;

    const root = switch (parsed.value) {
        .object => |obj| obj,
        else    => return null,
    };

    const response = root.get("response") orelse return null;
    return switch (response) {
        .string => |s| s,
        else    => null,
    };
}

fn trimSlash(s: []const u8) []const u8 {
    var end = s.len;
    while (end > 0 and s[end - 1] == '/') end -= 1;
    return s[0..end];
}

File: templates/rust/src/plugins/llamaindex.rs
//! LlamaIndex → Borgkit Plugin (Rust) — HTTP Bridge
//!
//! Wraps a LlamaIndex server so it participates in the Borgkit mesh as a
//! standard `IAgent`.  LlamaIndex agents and query engines can be served via
//! `llamaindex serve` or by using `llama_index.server` (FastAPI); this plugin
//! bridges both the `/chat` and `/query` endpoints.
//!
//! ── API contract ──────────────────────────────────────────────────────────────
//!
//!   Chat endpoint (default):
//!     POST {base_url}/chat
//!       Body:     { "message": "...", "chat_history": [] }
//!       Response: { "response": "...", "source_nodes": [...] }
//!
//!   Query endpoint (set invoke_route = "/query"):
//!     POST {base_url}/query
//!       Body:     { "query": "..." }
//!       Response: { "response": "...", "source_nodes": [...] }
//!
//! ── Setup ─────────────────────────────────────────────────────────────────────
//!
//!   Start a LlamaIndex server with:
//!     llamaindex serve --host 0.0.0.0 --port 8080
//!
//!   Or programmatically:
//!     from llama_index.server import LlamaIndexServer
//!     server = LlamaIndexServer(query_engine=my_engine)
//!     server.run()
//!
//! ── Usage ─────────────────────────────────────────────────────────────────────
//!
//!   use borgkit::plugins::llamaindex::{LlamaIndexPlugin, LlamaIndexService};
//!   use borgkit::plugins::base::PluginConfig;
//!
//!   let service = LlamaIndexService {
//!       base_url:     "http://localhost:8080".to_string(),
//!       invoke_route: "/chat".to_string(),
//!       capabilities: vec![("search".to_string(), "Search the knowledge base".to_string())],
//!       ..Default::default()
//!   };
//!
//!   let plugin = LlamaIndexPlugin::with_timeout(60);
//!   let agent  = plugin.wrap(service, PluginConfig {
//!       agent_id:     "borgkit://agent/llamaindex-rag".to_string(),
//!       owner:        "0xYourWallet".to_string(),
//!       network_host: "localhost".to_string(),
//!       network_port: 6174,
//!       ..Default::default()
//!   });

use async_trait::async_trait;
use serde_json::{json, Value};

use crate::plugins::base::{CapabilityDescriptor, BorgkitPlugin};
use crate::request::AgentRequest;
use crate::response::AgentResponse;

// ── Service config (the "agent" token) ───────────────────────────────────────

/// Configuration for a LlamaIndex HTTP server endpoint.
pub struct LlamaIndexService {
    /// Base URL, e.g. `http://localhost:8080`.
    pub base_url: String,

    /// POST path for invocation (default: `"/chat"`).
    ///
    /// Use `"/query"` to target the query engine endpoint instead.
    pub invoke_route: String,

    /// Optional Bearer token sent as `Authorization: Bearer <api_key>`.
    pub api_key: Option<String>,

    /// Explicit capability list: `(name, description)` pairs.
    ///
    /// When empty a single `"chat"` capability is synthesised automatically.
    pub capabilities: Vec<(String, String)>,
}

impl Default for LlamaIndexService {
    fn default() -> Self {
        Self {
            base_url:     "http://localhost:8080".to_string(),
            invoke_route: "/chat".to_string(),
            api_key:      None,
            capabilities: vec![],
        }
    }
}

// ── Plugin ────────────────────────────────────────────────────────────────────

/// HTTP bridge plugin for LlamaIndex servers.
pub struct LlamaIndexPlugin {
    client: reqwest::Client,
}

impl LlamaIndexPlugin {
    /// Create with a default 60-second timeout.
    pub fn new() -> Self {
        Self::with_timeout(60)
    }

    /// Create with a custom per-request timeout in seconds.
    pub fn with_timeout(secs: u64) -> Self {
        let client = reqwest::Client::builder()
            .timeout(std::time::Duration::from_secs(secs))
            .build()
            .unwrap_or_else(|_| reqwest::Client::new());
        Self { client }
    }
}

impl Default for LlamaIndexPlugin {
    fn default() -> Self { Self::new() }
}

// ── BorgkitPlugin impl ────────────────────────────────────────────────────────

#[async_trait]
impl BorgkitPlugin<LlamaIndexService> for LlamaIndexPlugin {
    fn extract_capabilities(&self, service: &LlamaIndexService) -> Vec<CapabilityDescriptor> {
        if service.capabilities.is_empty() {
            return vec![CapabilityDescriptor {
                name:           "chat".to_string(),
                description:    "Chat with or query the LlamaIndex agent".to_string(),
                input_schema:   None,
                output_schema:  None,
                price_per_call: None,
            }];
        }
        service.capabilities.iter().map(|(name, desc)| CapabilityDescriptor {
            name:           name.clone(),
            description:    desc.clone(),
            input_schema:   None,
            output_schema:  None,
            price_per_call: None,
        }).collect()
    }

    /// Build the LlamaIndex request body from an `AgentRequest`.
    ///
    /// For the `/chat` route the body is `{ "message": ..., "chat_history": [] }`.
    /// For the `/query` route the body is `{ "query": ... }`.
    /// The route is not available here so the sentinel field `"__borgkit_content"`
    /// is stored; final body assembly happens in `invoke_native`.
    fn translate_request(&self, request: &AgentRequest) -> Result<Value, String> {
        let content = request.payload.get("message")
            .or_else(|| request.payload.get("input"))
            .or_else(|| request.payload.get("query"))
            .map(|v| if v.is_string() {
                v.as_str().unwrap_or("").to_string()
            } else {
                v.to_string()
            })
            .unwrap_or_else(|| request.payload.to_string());

        Ok(json!({ "__borgkit_content": content }))
    }

    /// Extract the `response` field from the LlamaIndex response JSON.
    fn translate_response(&self, request_id: &str, output: Value) -> AgentResponse {
        let content = output
            .get("response")
            .map(|v| if v.is_string() {
                v.as_str().unwrap_or("").to_string()
            } else {
                v.to_string()
            })
            .unwrap_or_else(|| output.to_string());

        AgentResponse::success(
            request_id.to_string(),
            json!({ "content": content, "raw": output }),
        )
    }

    /// POST to `{base_url}{invoke_route}` with the appropriate body shape.
    ///
    /// Routes ending in `"/chat"` receive `{ "message": ..., "chat_history": [] }`;
    /// routes ending in `"/query"` receive `{ "query": ... }`.
    /// An optional Bearer token is added when `api_key` is configured.
    async fn invoke_native(
        &self,
        service: &LlamaIndexService,
        input:   Value,
    ) -> Result<Value, String> {
        let url = format!(
            "{}{}",
            service.base_url.trim_end_matches('/'),
            service.invoke_route,
        );

        let content = input
            .get("__borgkit_content")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();

        // Shape the body based on the configured route.
        let body = if service.invoke_route.ends_with("/query") {
            json!({ "query": content })
        } else {
            json!({ "message": content, "chat_history": [] })
        };

        let mut req = self.client.post(&url).json(&body);

        if let Some(ref key) = service.api_key {
            req = req.header("Authorization", format!("Bearer {key}"));
        }

        req.send()
            .await
            .map_err(|e| format!("LlamaIndex HTTP error: {e}"))?
            .json::<Value>()
            .await
            .map_err(|e| format!("LlamaIndex response parse error: {e}"))
    }
}
